setup.py

```diff
@@ -8,10 +18,127 @@
+    def run(self):
+        _setup_temp_egg_info(self)
+        self.run_command('static_bundles')
+        super().run()
+
+
+class BuildCommand(distutils.command.build.build):
+    def finalize_options(self):
+        if self.build_base == 'build':
+            # Use temporary directory instead, to avoid littering the source directory
+            # with a `build` sub-directory.
+            tempdir = tempfile.TemporaryDirectory()
+            self.build_base = tempdir.name
+            atexit.register(tempdir.cleanup)
+        super().finalize_options()
+
+    def run(self):
+        _setup_temp_egg_info(self)
+        self.run_command('static_bundles')
+        super().run()
+
+
+class DevelopCommand(setuptools.command.develop.develop):
+    def run(self):
+        self.run_command('static_bundles')
+        super().run()
+
+
+class StaticBundlesCommand(setuptools.command.build_py.build_py):
+
+    user_options = setuptools.command.build_py.build_py.user_options + [
+        ('bundle-type=', None,
+         'The bundle type. "min" (default) creates minified bundles,'
+         ' "dev" creates non-minified files.'),
+        ('skip-npm-reinstall', None,
+         'Skip running `npm install` if the `node_modules` directory already exists.'
+         ),
+        ('skip-rebuild', None,
+         'Skip rebuilding if the `sphinx_material` directory already exists.'),
+    ]
+
+    def initialize_options(self):
+        super().initialize_options()
+        self.bundle_type = 'min'
+        self.skip_npm_reinstall = None
+        self.skip_rebuild = None
+
+    def finalize_options(self):
+        super().finalize_options()
+        if self.bundle_type not in ['min', 'dev']:
+            raise RuntimeError('bundle-type has to be one of "min" or "dev"')
+
+        if self.skip_npm_reinstall is None:
+            self.skip_npm_reinstall = False
+
+        if self.skip_rebuild is None:
+            self.skip_rebuild = False
+
+    def run(self):
+        if self.skip_rebuild:
+            output_dir = os.path.join(package_root, 'static')
+            if os.path.exists(output_dir):
+                print('Skipping rebuild of package since %s already exists' %
+                      (output_dir, ))
+                return
+
+        target = {"min": "build", "dev": "build:dev"}
+
+        try:
+            t = target[self.bundle_type]
+            node_modules_path = os.path.join(root_dir, 'node_modules')
+            if (self.skip_npm_reinstall and os.path.exists(node_modules_path)):
+                print('Skipping `npm install` since %s already exists' %
+                      (node_modules_path, ))
+            else:
+                subprocess.call('npm i', shell=True, cwd=root_dir)
+            res = subprocess.call('npm run %s' % t, shell=True, cwd=root_dir)
+        except:
+            raise RuntimeError('Could not run \'npm run %s\'.' % t)
+
+        if res != 0:
+            raise RuntimeError('failed to build sphinx-material package')
```

Comment:
I just tried to integrate this modified branch into my project's _docs/requirements.txt_ like so
```
git+https://github.com/jbms/sphinx-material
sphinx-copybutton
```
But running `python -m pip install -r docs/requirements.txt` fails to build the sphinx-material theme. It's traceback ends on this line.

-----------------------------

Furthermore, I did a fresh install of npm and retried... It failed again, but this time, npm reported 
```
Error: Cannot find module 'fs/promises'
      at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
      at Function.Module._load (internal/modules/cjs/loader.js:562:25)
      at Module.require (internal/modules/cjs/loader.js:692:17)
      at require (internal/modules/cjs/helpers.js:25:18)
      at Object.<anonymous> (/tmp/pip-req-build-7y539i6_/tools/build/_/index.ts:24:1)
      at Module._compile (internal/modules/cjs/loader.js:778:30)
      at Module.m._compile (/tmp/pip-req-build-7y539i6_/node_modules/ts-node/src/index.ts:1056:23)
      at Module._extensions..js (internal/modules/cjs/loader.js:789:10)
      at Object.require.extensions.(anonymous function) [as .ts] (/tmp/pip-req-build-7y539i6_/node_modules/ts-node/src/index.ts:1059:12)
      at Module.load (internal/modules/cjs/loader.js:653:32)
```

is there a required npm (non-std) lib also? I did find [this commit comment](https://github.com/jbms/sphinx-material/commit/7f30d096a677c0e9a995e2617d8e9951821e8c66) to be helpful, but I'm not well versed with node.js.

Reply:
I think you may just need to update your version of nodejs.

Reply:
`nodejs -v` returns v10.19.0. I'll have to look into how to update that. I installed npm via (WSL) ubuntu apt, so I might find more problems from that.

I wondor if this requirement might break compatibility with readthedocs.org... I wouldn't think so since I can host docs generated with mkdocs-material.

Reply:
Could look into it more, I was just guessing, but I was testing with nodejs version 14 I believe.

mkdocs-material does not require users of the package to have node.js because the author re-computes the generated files and checks them into the repository after every change.

That has a number of drawbacks.  What I have setup instead in my branch is that if you build from the repository you will need nodejs because the generated file will be rebuilt automatically, but when you build a wheel (e.g. for publishing on pypi) the generated files will be pre-built.  So users will still be able to `pip install` without needing nodejs.

Reply:
ok I was able to update nodejs (from v10 to v16) via nvm, but I had to uninstall npm and nodejs via ubuntu apt first (it was problematic as I expected). I'm now able to install the the modified theme. Thank you for the assist!

I looked into readthedocs... [Their docker image source seems to install npm and nodejs via apt](https://github.com/readthedocs/readthedocs-docker-images/blob/1ebe82ab639cadb5c544491fa5fbc97f0b0049c5/Dockerfile#L125) 😦 which means this modified theme needs to be installed via pypi to avoid needing nodejs v14+. However, the RTFD manual suggests using a config file to ensure other software requirements are met, but I doubt that would help my situation (installing modified theme from git not pypi) due to the conflicts that installing nodejs from apt incur.